source/connect/connection-options/server-selection.txt

```diff
@@ -1,68 +1,193 @@
-   reporting its health, ``HeartbeatException`` contains the exception from the
-   last failed heartbeat. Analyzing the ``HeartbeatException`` on each
-   cluster node can assist in diagnosing most server selection issues.
-   The following heartbeat exceptions are common:
-   
-   * ``No connection could be made because the target machine actively
-     refused it``: The driver cannot see this cluster node. This can be
-     because the cluster node has crashed, a firewall is preventing
-     network traffic from reaching the cluster node or port, or some other
-     network error is preventing traffic from being successfully routed to
-     the cluster node.
-   * ``Attempted to read past the end of the stream``: This error
-     happens when the driver cannot connect to the cluster nodes due to a
-     network error, misconfigured firewall, or other network issue. To
-     address this exception, ensure that all cluster nodes are reachable.
-     This error commonly occurs when the client machine’s IP address is
-     not configured in the Atlas IPs Access List, which can be found under
-     the :guilabel:`Network Access` tab for your Atlas Project.
-   * ``The remote certificate is invalid according to the validation
-     procedure``: This error typically indicates a TLS/SSL-related problem
-     such as an expired/invalid certificate or an untrusted root CA. You
-     can use tools like ``openssl s_client`` to debug TLS/SSL-related
-     certificate problems.
\ No newline at end of file
+==========================
+Customize Server Selection
+==========================
+
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 1
+   :class: singlecol
+
+.. facet::
+   :name: genre
+   :values: reference
+
+.. meta::
+   :keywords: code example, read preference, write
+
+Overview
+--------
+
+All MongoDB drivers follow a defined algorithm when selecting a server to read or write
+from. By using the ``ClusterConfigurator`` property of a ``MongoClient`` object, you can
+customize this algorithm to choose the server that works best for your application.
+
+.. important::
+
+   Customizing the server-selection algorithm can have unintended consequences,
+   such as degraded read or write performance.
+
+Default Algorithm
+-----------------
+
+When the {+driver-short+} executes a read operation, it performs the following steps,
+in order, to select a MongoDB deployment:
+
+1. From the list of known servers, the {+driver-short+} selects all servers
+   that match the active read preference.
+
+#. If at least one readable server exists, the {+driver-short+} calls the user-defined
+   server-selector function and passes in the list from the previous step.
+
+#. The {+driver-short+} applies the ``LocalThreshold`` connection setting to the list of
+   servers returned from the function.
+
+#. The {+driver-short+} selects a server at random from the servers still on the list and
+   executes the operation against this server.
```

Comment:
S: Since we introduce this as "The driver performs the following steps..." it seems redundant to repeatedly say "The driver does XYZ" in the steps. It also adds wordiness to each step that I think can be simplified. I'd suggest simplifying to something like this
```suggestion
1. Selects all servers
   that match the active read preference from the list of known servers.

#. If at least one readable server exists, the driver calls the user-defined
   server-selector function and passes in the list from the previous step.

#. Applies the ``LocalThreshold`` connection setting to the list of
   servers returned from the function.

#. Selects a server at random from the servers still on the list and
   executes the operation against this server.
```